shell/platform/android/io/flutter/plugin/editing/SpellCheckPlugin.java

```diff
@@ -0,0 +1,134 @@
+   * io.flutter.embedding.engine.systemchannels.SpellCheckChannel}, and closes the most recently
+   * opened {@code SpellCheckerSession}.
+   *
+   * <p>Do not invoke any methods on a {@code SpellCheckPlugin} after invoking this method.
+   */
+  public void destroy() {
+    mSpellCheckChannel.setSpellCheckMethodHandler(null);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+  }
+
+  @Override
+  public void initiateSpellCheck(@NonNull String locale, @NonNull String text) {
+    performSpellCheck(locale, text);
+  }
+
+  /** Calls on the Android spell check API to spell check specified text. */
+  public void performSpellCheck(@NonNull String locale, @NonNull String text) {
+    String[] localeCodes = locale.split("-");
+    Locale localeFromString = LocalizationPlugin.localeFromString(locale);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+    mSpellCheckerSession =
+        mTextServicesManager.newSpellCheckerSession(
+            null,
+            localeFromString,
+            this,
+            /** referToSpellCheckerLanguageSettings= */
+            true);
+
+    TextInfo[] textInfos = new TextInfo[] {new TextInfo(text)};
+    mSpellCheckerSession.getSentenceSuggestions(textInfos, MAX_SPELL_CHECK_SUGGESTIONS);
+  }
+
+  /**
+   * Callback for Android spell check API that decomposes results and send results through the
+   * {@link SpellCheckChannel}.
+   *
+   * <p>Spell check results will be encoded as a string representing the span of that result, with
+   * the format [start_index.end_index.suggestion_1,suggestion_2,suggestion_3], where there may be
+   * up to 5 suggestions.
+   */
+  @Override
+  public void onGetSentenceSuggestions(SentenceSuggestionsInfo[] results) {
+    ArrayList<String> spellCheckerSuggestionSpans = new ArrayList<String>();
+
+    if (results.length > 0) {
+      SentenceSuggestionsInfo spellCheckResults = results[0];
+
+      for (int i = 0; i < spellCheckResults.getSuggestionsCount(); i++) {
+        SuggestionsInfo suggestionsInfo = spellCheckResults.getSuggestionsInfoAt(i);
+        int suggestionsCount = suggestionsInfo.getSuggestionsCount();
+
+        if (suggestionsCount > 0) {
+          String spellCheckerSuggestionSpan = "";
+          int start = spellCheckResults.getOffsetAt(i);
+          int length = spellCheckResults.getLengthAt(i);
+
+          spellCheckerSuggestionSpan += (String.valueOf(start) + ".");
+          spellCheckerSuggestionSpan += (String.valueOf(start + (length - 1)) + ".");
+
+          for (int j = 0; j < suggestionsCount; j++) {
+            spellCheckerSuggestionSpan += (suggestionsInfo.getSuggestionAt(j) + "/n");
+          }
+
+          spellCheckerSuggestionSpans.add(
+              spellCheckerSuggestionSpan.substring(0, spellCheckerSuggestionSpan.length() - 2));
+        }
+      }
+    }
+
+    mSpellCheckChannel.updateSpellCheckResults(spellCheckerSuggestionSpans);
```

Comment:
Similar to my previous comment, if the framework calls initiateSpellCheck multiple times before receiving updateSpellCheckResults, what guarantees does it have about the order of incoming updateSpellCheckResults calls?

Reply:
I've seen the results consistently come back in order, and like I said in the previous comment, in cases where it takes a bit longer to get spell check results back, the older calls get consumed, so it seems like the spell check service handles this already.

Reply:
This is pretty theoretical but do you think this is a valid problem?

  1. The framework calls initiateSpellCheck with string "recieve".
  2. The engine receives this and calls updateSpellCheckResults.
  3. The user has quickly pasted something new, so framework calls initiateSpellCheck with string "thier".
  4. The engine is called back on updateSpellCheckResults with the suggestions for "recieve" and sends them to the framework.
  5. The framework receives the suggestions for "recieve".  The last call it made to initiateSpellCheck was for string "thier", and that's what's in the text field, so it incorrectly applies the suggestions for "recieve" to "thier".

In reality, I think this would only be a problem for a few milliseconds because the framework would then receive the correct suggestions for "thier".  These kinds of problems are rare in reality but hard to eliminate because everything is asynchronous.  We have similar problems with updateTextEditingState itself, and we currently don't do anything about them, so maybe this is fine here too.

One thing we could do, though, is to send the text (or a version or a hash) along with the suggestions so we could make sure we apply them to the correct text.  I don't know if that is worth it here.

@LongCatIsLooong I know you've worked with these asynchronous problems before, any thoughts here?  I'm probably being too nitpicky but I want a second opinion to tell me not to worry about this.

Reply:
> In reality, I think this would only be a problem for a few milliseconds because the framework would then receive the correct suggestions for "thier".

I was suggesting to send the string from the request back in the response, so the framework can handle that case. I think it would be great to address this issue.

Reply:
The spell checking process probably should not be initiated before the text input formatters are applied, the `onChanged` callback is called, and the `TextEditingController` listeners are notified in the framework (as we allow developers to change the text in these steps), to minimize the chance of this kind of races happening. But if the `initiateSpellCheck` API is exposed to the developer then I think we should provide a way to allow them to tell if the result is still relevant.

I'm ok with including the original string from the request but that could increase the payload size by a lot if the developer sends over the entire Bible for spellchecking. From the framework's perspective ideally the spell check service should take a `String` and return a `Future<ListOfSuggestions>`, with this kind of one-to-one correspondence the text field can determine whether the suggestion results are still relevant, and discards the results if the text has already changed.

Reply:
> I'm ok with including the original string from the request but that could increase the payload size by a lot if the developer sends over the entire Bible for spellchecking. From the framework's perspective ideally the spell check service should take a `String` and return a `Future<ListOfSuggestions>`, with this kind of one-to-one correspondence the text field can determine whether the suggestion results are still relevant, and discards the results if the text has already changed.

I discussed a similar approach with @blasten yesterday and I agree that this is the best way to proceed. Even if I do use the relevant text to ensure the results are still relevant, I don't have to send them to the engine and back, as I already found a way to recall the `TextEditingValue` that the spell check results are related to (by passing them [here](f637dd3f4492c05eb332ccb7e907b8e3b7413ced7d6f2fb125f58e98e9db0ceaR52)). 

Down the line, I think exploring techniques for speeding up the entire process like only spell checking sentences that have changed versus the entirety of the inputted text could help, as well, as I suspect that the case @justinmc described would occur more frequently as the length of the text being spell checked increases. I definitely think that case should be addressed before any framework PRs land, but I believe it shouldn't block this PR since potential fixes would be in the framework.

Reply:
Ah thanks for the clarification! Looking at the framework-side implementation, it seems to be expecting the spellchecking service to guarantee the requests are replied to in FIFO order, but the framework-side API actually doesn't provide the same FIFO guarantee its implementation relies on, to the API consumer (because the signature of the function is `(Locale, TextEditingValue) -> Future<List<SpellCheckerSuggestionSpan>>`).

For Android and iOS builtin spellchecking services it's easy to satisfy the FIFO requirement but I think this can be a bit too "strong" of a requirement in some other occasions. If someone wants to implement a spellchecking service that, say, fetches results from a web server, it could be a hassle to guarantee the FIFO ordering.

Reply:
I do agree it could be hassle, but I do think it could be largely addressed on the framework side. Due to the fact that the framework may attempt to build the `TextSpanTree` with spell check results that are not up to date, I added some logic there that could do some corrections on those results to build the `TextSpanTree` as accurate as it can guess. This may address that problem to a large degree, but of course, I acknowledge that it would be best if we could guarantee the FIFO ordering.